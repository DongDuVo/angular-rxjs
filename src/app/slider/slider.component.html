<div class="slider" #slider>
  <div #button
    class="slider__button"
    (mousedown)="mouseDown$.next($event)"
    [ngStyle]="buttonStyle$ | async"
  ></div>
</div>
