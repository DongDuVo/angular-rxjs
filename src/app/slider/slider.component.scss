.slider {
  @apply border-pink-800;
  @apply bg-white;
  @apply rounded-full;
  @apply border;
  @apply relative;
  @apply w-64;
  @apply h-12;
  padding: 1px;

  &__button {
    @apply bg-purple-700;
    @apply border-pink-500;
    @apply text-white;
    @apply border;
    @apply rounded-full;
    @apply flex;
    @apply justify-center;
    @apply items-center;
    @apply absolute;
    @apply cursor-pointer;

    left: 1px;
    width: 44px;
    height: 44px;
  }
}
