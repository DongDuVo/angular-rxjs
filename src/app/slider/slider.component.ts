import { ChangeDetectionStrategy, Component, ElementRef, ViewChild, ViewEncapsulation } from '@angular/core';
import { animationFrameScheduler, defer, fromEvent, of, Subject, merge, generate } from 'rxjs';
import { delay, map, switchMapTo, takeUntil, throttleTime, withLatestFrom } from 'rxjs/operators';

@Component({
  selector: 'app-slider',
  templateUrl: './slider.component.html',
  styleUrls: ['./slider.component.scss'],
  encapsulation: ViewEncapsulation.None,
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class SliderComponent {
  @ViewChild('slider', { static: true, read: ElementRef })
  slider!: ElementRef<HTMLDivElement>;

  @ViewChild('button', { static: true, read: ElementRef })
  button!: ElementRef<HTMLDivElement>;

  curPos = 1;

  mouseDown$ = new Subject<MouseEvent>();
  mouseUp$ = new Subject<number>();
  buttonStyle$ = merge(this.mouseDown$.pipe(
      switchMapTo(
        fromEvent<MouseEvent>(document, 'mousemove').pipe(
          takeUntil(fromEvent(document, 'mouseup')),
          throttleTime(0, animationFrameScheduler),
          withLatestFrom(defer(() => of(this.slider.nativeElement.clientWidth))),
          map(([moveEvent, sliderWidth]) => {
            const position = moveEvent.clientX + 1 - 44;
            this.curPos = position <= 1 ? 1 : Math.min(sliderWidth - 1 - 44, position);
            return { 'left.px': this.curPos };
          })
        )
      )
    ), this.mouseUp$.pipe(
        switchMapTo(
          generate(this.curPos, x => x >= 1, x => x - 1, x => x).pipe(
            delay(50),
            map((pos) => {
              console.log('pos ', pos)
              return { 'left.px': pos };
            }
          )
        )
      )
    )
  )
}
