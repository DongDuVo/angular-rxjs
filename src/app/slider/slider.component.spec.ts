import { NO_ERRORS_SCHEMA } from '@angular/core';
import { RouterTestingModule } from '@angular/router/testing';
import { render, RenderResult } from '@testing-library/angular';

import { SliderComponent } from './slider.component';

describe('MainComponent', () => {
  let component: RenderResult<SliderComponent>;

  beforeEach(async () => {
    component = await render(SliderComponent, {
      imports: [RouterTestingModule],
    });
  });

  it('should match snapshot', () => {
    expect(component.container).toMatchSnapshot();
  });
});
