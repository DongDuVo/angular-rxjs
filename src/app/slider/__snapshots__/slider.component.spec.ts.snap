// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainComponent should match snapshot 1`] = `
<div
  id="root0"
>
  <div
    class="slider"
  >
    <div
      class="slider__button"
    />
  </div>
</div>
`;
